Honor EdgeZero app config store default (#879)
IABTechLab/trusted-server: docs/guide/configuration.md

EdgeZero entry point. The former `edgezero_enabled` and `edgezero_rollout_pct`
canary keys are no longer read.

The Fastly service must provide the logical config store selected by
`[stores.config].default` in `edgezero.toml`. By default, the logical store ID
is also the platform store name and the app-config blob key. Deployments can
override those independently with
`EDGEZERO__STORES__CONFIG__<ID>__NAME` and

> `EDGEZERO__STORES__CONFIG__<ID>__KEY`.

Reviewer: 🔧 **wrench** — The `NAME` half of this sentence is correct; the `KEY` half leads an operator into the fail-closed startup described three lines below.

Only `NAME` round-trips. Probed with this branch's own CLI:

```
# NAME override — honored by push
$ EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__NAME=review_probe_store \
    ts config push --adapter fastly --local --dry-run
[edgezero] ... -> store `trusted_server_config` (platform name `review_probe_store`)

# KEY override — silently ignored by push
$ EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__KEY=review_probe_key \
    ts config push --adapter fastly --local --dry-run
# no remote at key `trusted_server_config`; all leaves added
```

**Cause**: `edgezero-cli` v0.0.4 resolves the blob key as `args.key.clone().unwrap_or_else(|| ctx.store.logical.clone())` (`edgezero-cli/src/config.rs:300-303`) and never calls `EnvConfig::store_key` — `store_key` appears nowhere in that crate. The runtime `default_config_key()` (`settings_data.rs:42-48`) *does* consult it.

**Failure scenario**: operator follows this paragraph, exports `EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__KEY=staging`, runs `ts config push --adapter fastly`, sees the push succeed. The blob lands at key `trusted_server_config`; the app reads key `staging`, finds nothing, and fails closed at startup — exactly the outcome documented at line 1479 for an absent entry.

The asymmetry predates this PR, but this PR is what documents the override as supported, and `config_defaults_honor_edgezero_store_name_and_key_overrides` asserts the runtime half only — which reads as a guarantee it isn't.

**Fix** — either drop `KEY` from the sentence, or pair it with the flag that makes it work:

> The platform store name is overridable with `EDGEZERO__STORES__CONFIG__<ID>__NAME`, honored by both the runtime and `ts config push`. The blob key is not symmetric: the runtime reads `EDGEZERO__STORES__CONFIG__<ID>__KEY`, but `ts config push` does not — pass `--key` to match it.
>
> ```bash
> EDGEZERO__STORES__CONFIG__<ID>__KEY=staging ts config push --adapter fastly --key staging
> ```

Author: Addressed in 2760e429: documentation now distinguishes the supported NAME round trip from the runtime-only KEY override and shows the matching ts config push --key command.

The resolved store and key must contain a valid Trusted Server app-config blob

> envelope. An absent or empty entry makes application startup fail closed. Use

Reviewer: 👍 **praise** — This replaces a claim that was simply wrong ("The store may be empty unless another feature adds keys to it") with the actual behavior. Worth calling out because the old text would have led an operator to skip seeding the blob entirely.

Both new commands also check out against `edgezero-cli` v0.0.4: `ProvisionArgs::adapter` is `required = true`, and `ConfigPushArgs::local` is documented there as editing `[local_server.config_stores]` in the adapter's `fastly.toml` for Viceroy to read.

Author: Thanks. Kept the corrected fail-closed startup guidance and verified the documentation formatter.

EdgeZero entry point. The former `edgezero_enabled` and `edgezero_rollout_pct`
canary keys are no longer read.

The Fastly service must provide the logical config store selected by
`[stores.config].default` in `edgezero.toml`. By default, the logical store ID
is also the platform store name and the app-config blob key. Deployments can
override the platform store name with

> `EDGEZERO__STORES__CONFIG__<ID>__NAME`; both the runtime and `ts config push`

Reviewer: ❓ **question** — "both the runtime and `ts config push` honor that override" — is the runtime half true for Fastly? This paragraph is under **Fastly Runtime Config Store**.

The push half I verified (`edgezero-cli/src/config.rs:435`). The runtime half conflicts with the repo's own Fastly deployment guidance in `.claude/skills/deploying-trusted-server-to-fastly/SKILL.md:28-34`, which documents a different mechanism entirely: create a service-scoped physical store, `fastly resource-link create … --name trusted_server_config`, and then "at runtime the service opens `trusted_server_config` and gets the linked physical store." Under that model `NAME` is a push-side knob and the resource link does the runtime mapping — `default_config_store_name()` in the guest would still return the manifest default.

I could not find anything in this repo that puts a custom `EDGEZERO__…` variable into a Fastly Compute guest's environment, so I do not want to assert either way.

If `NAME`/`KEY` are only honored at runtime where `std::env` is actually populated (Axum dev, Viceroy), both sentences want scoping to those adapters, and this Fastly section should point at the resource-link approach instead.

Author: Addressed in 391fbb52. The Fastly guide now states that the guest opens the manifest logical store and Fastly resource links map it to the physical store. `NAME` is documented only as the `ts config push` target override.

honor that override. The runtime also reads
`EDGEZERO__STORES__CONFIG__<ID>__KEY`, but `ts config push` requires a matching
`--key` argument to publish the blob at that key:

```bash

> EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__KEY=staging \

Reviewer: 🔧 **wrench** — This example sets the runtime env var on the `ts config push` process, which never reads it.

I checked the pinned CLI (`edgezero-cli/src/config.rs:435` and `:1035`): the push path resolves the store with `env_config.store_name("config", &logical)` and never calls `store_key`. That is exactly why `--key` is mandatory here. The `EDGEZERO__STORES__CONFIG__…__KEY` prefix on this command changes nothing about the push, and it is not set anywhere the runtime will see it.

An operator who copies this block ends up with the blob published at `staging` and a runtime still resolving `trusted_server_config` — the fail-closed startup described three lines below. That is the same trap the last round flagged, now in copy-pasteable form.

**Fix** — the push command needs only the flag:

```bash
ts config push --adapter fastly --key staging
```

If the runtime half is reachable on this adapter (see my question on line 1475), show it as a separate step in whatever place actually sets the runtime environment, so the two are not read as one command.

Author: Addressed in 391fbb52. Removed the runtime `KEY` override and `staging` example from the Fastly guide. The documented flow now keeps the default blob key and uses a resource link for runtime mapping, so the push and read cannot diverge.

EdgeZero entry point. The former `edgezero_enabled` and `edgezero_rollout_pct`
canary keys are no longer read.

The Fastly service opens the logical config store selected by
`[stores.config].default` in `edgezero.toml` and reads the app-config blob at
its default key. A Fastly resource link maps that logical store name to the
physical config store for the service at runtime.

For a service-specific physical store, create the store, link it with the
logical name, then push to that physical store:

```bash
fastly config-store create --name <physical-store-name>
fastly resource-link create --service-id <service-id> --version <version> \
  --resource-id <config-store-id> --name trusted_server_config
EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__NAME=<physical-store-name> \

>   ts config push --adapter fastly

Reviewer: 🔧 **wrench** — This sequence cannot be run as written on a service that has ever been deployed, and if it does succeed it may never take effect.

`fastly resource-link create --version <version>` requires an *editable* service version. Confirmed against the installed CLI: `--version` accepts `'latest'`, `'active'`, `'staged'`, or a version number, and `--autoclone` is the flag that makes it usable — "If the selected service version is not editable, clone it and use the clone."

EdgeZero's own Fastly provisioner emits exactly that form as its remediation string when it detects an already-deployed service (`edgezero-adapter-fastly/src/cli.rs:831`), including the caveat this block is missing:

```
fastly resource-link create --service-id=<SID> --resource-id=<STORE-ID> --version=latest --autoclone --name=<name>
(the link clones the active version so existing traffic is not affected until you `fastly service-version activate`)
```

**Failure scenario**: an operator on a live service follows this block. With `--version active` (or the active version's number) the command fails outright — the version is locked. If they instead pass a draft version number it succeeds, but the link exists only on that draft: the running service keeps resolving whatever store was linked before, or the shared account-level default. Nothing tells them to activate, so that second path fails *silently* — and it lands them in precisely the cross-service config overwrite this section exists to prevent.

**Fix** — add `--autoclone`, and activate after the config is pushed so the new version never goes live against an unseeded store:

```suggestion
fastly resource-link create --service-id <service-id> --version latest --autoclone \
  --resource-id <config-store-id> --name trusted_server_config
EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__NAME=<physical-store-name> \
  ts config push --adapter fastly
fastly service-version activate --service-id <service-id> --version latest
```

_(verified against the pinned `docs/` Prettier — `guide/configuration.md` still passes `--check` with these bytes applied)_

Author: Addressed in 953f0df2. The guide now separates first-deployment and already-deployed flows. The existing-service sequence uses `--version latest --autoclone`, verifies the target with `ts config push --dry-run`, seeds the physical store, and activates the cloned version only afterward.

```

The resource-link name must match the logical store ID, and the pushed store
must contain a valid Trusted Server app-config blob envelope at its default key.
An absent or empty entry makes application startup fail closed. Use the Trusted

> Server CLI to provision the store and publish the validated config.

Reviewer: 🤔 **thinking** — These paragraphs restate `.claude/skills/deploying-trusted-server-to-fastly/SKILL.md:22-35` (create a service-scoped physical store → link it under the logical name → push with the `NAME` override) but drop the two guardrails that file attaches to the same procedure:

- "**Always `ts config push --dry-run` first.** It prints the exact target physical store; confirm it is the service-scoped one, not a shared default, before writing."
- "**Always pass `--service-id=<SID>`** to `fastly compute publish`. A checked-in `fastly.toml` may pin a different (or dead) `service_id`."

Both exist because the failure mode is silently overwriting another service's config, which is the exact hazard this section is introducing the procedure to avoid. Two copies of the procedure will drift, and the copy without the guardrails is the one in the published guide. Either point at the skill from here, or move the guardrails into the guide and let the skill reference it.

Separately, the two flows below read as alternatives without saying which to pick. Line 1490 says "Use the Trusted Server CLI to provision the store", and lines 1498-1502 give `ts provision --adapter fastly` as the production path — but `ts provision` creates the store under the *logical* default name: `store.platform` falls back to the logical id when no `NAME` override is set (`edgezero-adapter-fastly/src/cli.rs:236-262`), so it produces `fastly config-store create --name=trusted_server_config`. That is the shared account-level default the block above deliberately sidesteps. Worth stating explicitly whether `ts provision` is the single-service path and the manual sequence is the multi-service one, since on a shared account picking the wrong one is the collision case.

Author: Addressed in 953f0df2. The published guide now distinguishes the default first-deployment path, the service-specific first-deployment path for shared accounts, and the explicit resource-link path for already-deployed services. Each production path includes a dry run and tells operators to confirm the physical target. The guide does not link to the internal `.claude` skill. It also does not show `fastly compute publish`; every service-version command it does show passes `--service-id` explicitly.

EdgeZero entry point. The former `edgezero_enabled` and `edgezero_rollout_pct`
canary keys are no longer read.

The Fastly service opens the logical config store selected by
`[stores.config].default` in `edgezero.toml` and reads the app-config blob at
its default key. A Fastly resource link maps that logical store name to the
physical config store for the service at runtime.

Fastly store names are account-level. For a first deployment, the default setup
is safe only when no other service in the account uses the
`trusted_server_config` physical store:

```bash
ts provision --adapter fastly
ts config push --adapter fastly --dry-run
ts config push --adapter fastly
```

> For the first deployment of a service in a shared account, select a
> service-specific physical store during both provisioning and config push. The
> first deployment applies the generated setup entry and links the physical store
> under the logical name:
> 
> ```bash
> EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__NAME=<physical-store-name> \
>   ts provision --adapter fastly
> EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__NAME=<physical-store-name> \
>   ts config push --adapter fastly --dry-run
> EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__NAME=<physical-store-name> \
>   ts config push --adapter fastly
> ```
> 
> For an already-deployed service, Fastly does not reapply setup entries. Create
> and link the service-specific physical store explicitly, seed it, then activate
> the cloned service version:

Reviewer: 🔧 **wrench** — This sequence links the service-scoped store under its *physical* name, but the Fastly entry point opens the *logical* one. The service fails to boot.

`ts provision` keys its generated Fastly setup table on the env-resolved platform name, not on the logical id. Reproduced locally against this head with a fake override:

```
$ EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__NAME=pub_a_ts_config \
    ts provision --adapter fastly --dry-run
would run `fastly config-store create --name=pub_a_ts_config` and append
  [setup.config_stores.pub_a_ts_config] to ./fastly.toml (logical id `trusted_server_config`)
```

That matches the source: `edgezero-adapter-fastly/src/cli.rs:227-244` binds `name = store.platform.as_str()` and passes it to both `create_fastly_store` and `append_fastly_setup`, which writes `[setup.<kind>_stores.<name>]`. Fastly consumes that table key as the resource-link name on the first `compute publish`, so the link is called `pub_a_ts_config`.

The runtime never asks for that name. `main.rs:52` opens `default_config_store_name()`, which is `EnvConfig::from_env().store_name("config", "trusted_server_config")` — and Compute has no process environment for operator-set variables, so it resolves to the bare logical id. `EdgeZeroFastlyConfigStore::try_open` (`edgezero-adapter-fastly/src/config_store.rs:50`) forwards that string to Fastly verbatim; there is no logical-to-physical mapping inside.

**Failure scenario**: an operator adding a second service to a shared account follows this block. Provision creates `pub_a_ts_config` and the setup entry; `ts config push` writes a valid envelope into it; the first `fastly compute publish` links it as `pub_a_ts_config`. The service then opens `trusted_server_config`, finds no linked store, and fails closed at startup — with the config sitting correctly seeded in a store the binary will never look at.

The section already states the right rule two paragraphs down: *"The resource-link name must match the logical store ID."* `.claude/skills/deploying-trusted-server-to-fastly/SKILL.md:27-29` prescribes the same `--name trusted_server_config`. Only this block departs from it.

**Fix** — drop the provision-based variant and let the explicit create-and-link block below cover the service-scoped case for both first and subsequent deployments, since it is the only sequence that produces a link named `trusted_server_config`:

```suggestion
For a service-specific physical store, create and link it explicitly. Do not use
`ts provision` for this: it keys its generated `[setup.config_stores.<name>]`
entry on the _physical_ store name, so the resource link a first deployment
creates from that entry is named `<physical-store-name>` — while the entry point
opens `trusted_server_config` and never finds it. Fastly also skips `[setup]`
entirely on a service that has already been deployed. Create the store, link it
under the logical name, seed it, then activate the cloned service version:
```

_(verified against the pinned `docs/` Prettier — `guide/configuration.md` still passes `--check` with these bytes applied, and the pre/post-verification patch snapshots are byte-identical)_

Author: Addressed in e2637d7a5. Removed the unsupported `ts provision` plus `__NAME` first-deployment sequence. The guide now explains that the pinned provisioner uses the physical name as the setup-table and resource-link name, and directs shared-account services through an explicit `trusted_server_config` resource link that is seeded before activation.

first. Then create and link the service-specific physical store explicitly,
seed it, and activate the linked version before publishing traffic:

```bash
fastly config-store create --name <physical-store-name>
fastly resource-link create --service-id <service-id> --version latest --autoclone \
  --resource-id <config-store-id> --name trusted_server_config
EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__NAME=<physical-store-name> \
  ts config push --adapter fastly --dry-run
EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__NAME=<physical-store-name> \
  ts config push --adapter fastly
fastly service-version activate --service-id <service-id> --version latest
```

Confirm that each dry run names the intended physical store before writing. In
a shared account, it must be the service-specific store rather than the
account-level default. The resource-link name must match the logical store ID,
and the physical store must contain a valid Trusted Server app-config blob
envelope at its default key. An absent or empty entry makes application startup
fail closed.

**Local development** (writes the entry used by Viceroy in `fastly.toml`):

```bash
ts config push --adapter fastly --local

> ```

Reviewer: 🤔 **thinking** — Two footguns land on this one command, and the section above sets up the first of them.

`--local` writes to `[local_server.config_stores.<physical-name>]`, where `<physical-name>` is the env-resolved store name — but the Viceroy guest opens the *logical* `trusted_server_config`. An operator who exported `EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__NAME` for the shared-account block a few lines up still has it in their shell here, so the blob lands in a block nothing reads and startup fails with a confusing missing-key error.

Second, this command writes the full app-config blob — `proxy_secret`, the EC passphrase, and the admin password among it — into `fastly.toml`, which is tracked. That is pre-existing behaviour (`docs/guide/auction-orchestration.md:848` documents the same command without a warning), so one line here rather than a rework:

`````suggestion
**Local development** (writes the entry used by Viceroy in `fastly.toml`):

Unset any `EDGEZERO__STORES__CONFIG__TRUSTED_SERVER_CONFIG__NAME` override
first: `--local` writes to `[local_server.config_stores.<physical-name>]`, while
the Viceroy guest opens the logical `trusted_server_config` store. This command
also writes the full app-config blob — including `proxy_secret`, the EC
passphrase, and the admin password — into the tracked `fastly.toml`; do not
commit the result.

```bash
ts config push --adapter fastly --local
```
`````

_(Verified in a scratch worktree at this head: the pinned `docs` Prettier reports `All matched files use Prettier code style!`. The outer fence is widened to five backticks so the nested ```` ```bash ```` block does not close it early.)_
